File: RotateArray/LeftRotateByNPlaces.ts
/**
 * Brute force: Rotate array left by N places by rotating 1 place N times.
 *
 * Idea:
 * - For each rotation:
 *   - Store arr[0]
 *   - Shift all elements left by 1
 *   - Put stored element at arr[n-1]
 *
 * Time Complexity: O(n * k)
 * Space Complexity: O(1)
 *
 * Edge cases:
 * - null/undefined -> error
 * - [] or [x] -> no rotation
 * - k > length -> use k = k % length
 */
export function leftRotateBrute(arr: number[], k: number): void {
    if (!arr) {
        throw new Error("Array must be non-null");
    }

    const n = arr.length;
    if (n <= 1) return;

    k = k % n;
    if (k === 0) return;

    for (let rot = 0; rot < k; rot++) {
        const first = arr[0];

        for (let i = 1; i < n; i++) {
            arr[i - 1] = arr[i];
        }

        arr[n - 1] = first;
    }
}

/**
 * Optimised rotation using Reversal Algorithm.
 *
 * Idea:
 * - Reverse first k elements
 * - Reverse last n-k elements
 * - Reverse whole array
 *
 * Time Complexity: O(n)
 * Space Complexity: O(1)
 */
export function leftRotateOptimised(arr: number[], k: number): void {
    if (!arr) {
        throw new Error("Array must be non-null");
    }

    const n = arr.length;
    if (n <= 1) return;

    k = k % n;
    if (k === 0) return;

    reverse(arr, 0, k - 1);
    reverse(arr, k, n - 1);
    reverse(arr, 0, n - 1);
}

function reverse(arr: number[], l: number, r: number): void {
    while (l < r) {
        const temp = arr[l];
        arr[l] = arr[r];
        arr[r] = temp;
        l++;
        r--;
    }
}

/** Utility printer */
function print(arr: number[]): void {
    console.log("[" + arr.join(", ") + "]");
}

/** Test Cases */
const tests: number[][] = [
    [1, 2, 3, 4, 5],
    [10],
    [],
    [-1, -2, -3, -4, -5],
    [5, 5, 5, 5],
    [1, 2]
];

const ks = [2, 3, 4, 7, 1, 5];

for (let i = 0; i < tests.length; i++) {
    console.log("Original:", tests[i], "Rotate by:", ks[i]);

    let b = [...tests[i]];
    let o = [...tests[i]];

    leftRotateBrute(b, ks[i]);
    console.log("Brute:     ", b);

    leftRotateOptimised(o, ks[i]);
    console.log("Optimised: ", o);

    console.log();
}

File: RotateArray/LeftRotateByNPlaces.java
public class LeftRotateByNPlaces {

    /**
     * Brute force left rotation by N places.
     *
     * Idea:
     * - Perform left rotation by 1 place, N times.
     * - Each rotation shifts all elements left once.
     *
     * Time Complexity: O(n * k)
     * - n = array length
     * - k = rotation count
     * - Because each of the k rotations shifts n elements.
     *
     * Space Complexity: O(1)
     * - Only temporary variable used.
     *
     * Edge Cases:
     * - null array -> IllegalArgumentException
     * - [] → do nothing
     * - n = 1 → no rotation
     * - k = 0 → no rotation
     * - k > length → use k = k % length
     */
    public static void leftRotateBrute(int[] arr, int k) {
        if (arr == null) {
            throw new IllegalArgumentException("Array must be non-null");
        }

        int n = arr.length;
        if (n <= 1) return;

        k = k % n;
        if (k == 0) return;

        for (int r = 0; r < k; r++) {
            int first = arr[0];
            for (int i = 1; i < n; i++) {
                arr[i - 1] = arr[i];
            }
            arr[n - 1] = first;
        }
    }

    /**
     * Optimised left rotation by N places using the Reversal Algorithm.
     *
     * Idea:
     * To rotate array left by k:
     *   1. Reverse first k elements.
     *   2. Reverse remaining (n - k) elements.
     *   3. Reverse the entire array.
     *
     * Example:
     *   arr = [1,2,3,4,5], k = 2
     *   Step1 reverse(0..1) → [2,1,3,4,5]
     *   Step2 reverse(2..4) → [2,1,5,4,3]
     *   Step3 reverse whole → [3,4,5,1,2]
     *
     * Time Complexity: O(n)
     * Space Complexity: O(1)
     *
     * Edge Cases:
     * - null array -> IllegalArgumentException
     * - [] or [x] → no rotation
     * - k = 0 → no rotation
     * - k > length → use k = k % length
     */
    public static void leftRotateOptimised(int[] arr, int k) {
        if (arr == null) {
            throw new IllegalArgumentException("Array must be non-null");
        }

        int n = arr.length;
        if (n <= 1) return;

        k = k % n;
        if (k == 0) return;

        reverse(arr, 0, k - 1);
        reverse(arr, k, n - 1);
        reverse(arr, 0, n - 1);
    }

    /** Helper: reverse a subarray from l to r */
    private static void reverse(int[] arr, int l, int r) {
        while (l < r) {
            int temp = arr[l];
            arr[l] = arr[r];
            arr[r] = temp;
            l++;
            r--;
        }
    }

    /** Utility to print array */
    private static void printArray(int[] arr) {
        System.out.print("[");
        for (int i = 0; i < arr.length; i++) {
            System.out.print(arr[i]);
            if (i < arr.length - 1) System.out.print(", ");
        }
        System.out.println("]");
    }

    public static void main(String[] args) {

        int[][] testArrays = {
            {1,2,3,4,5},
            {10},
            {},
            {-1,-2,-3,-4,-5},
            {5,5,5,5},
            {1,2}
        };

        int[] rotations = {2, 3, 4, 7, 1, 5};

        for (int i = 0; i < testArrays.length; i++) {
            int[] arr1 = testArrays[i].clone();
            int[] arr2 = testArrays[i].clone();
            int k = rotations[i];

            System.out.println("Original:");
            printArray(arr1);
            System.out.println("Rotate by " + k);

            leftRotateBrute(arr1, k);
            System.out.print("Brute Result:     ");
            printArray(arr1);

            leftRotateOptimised(arr2, k);
            System.out.print("Optimised Result: ");
            printArray(arr2);

            System.out.println();
        }
    }
}
